File: laba6/C#/Expression/Expression.cs
using System;

namespace Expression
{
    public class TaskExpression
    {
        private double _variable1, _variable2, _variable3, _variable4;  // a,b,c,d
        public TaskExpression(double a, double b, double c, double d) // Constructor with variables for expression
        {
            _variable1 = a;
            if (d != 0)
            {
                if (4 * b - c > 0)
                {
                    if (b + c / d - 1 != 0)
                    {
                        _variable2 = b;
                        _variable3 = c;
                        _variable4 = d;
                    }
                    else { throw new ArithmeticException($"Wrong value(s) of b/c/d. b + c / d - 1 = {b + c / d - 1}"); }
                }
                else { throw new ArithmeticException($"Wrong value(s) of b and/or c. 4 * b - c = {4 * b - c}"); }
            }
            else { throw new DivideByZeroException("Wrong value of d"); }
        }

        public double GetA() => _variable1;  // Get value of A
        public double GetB() => _variable2;  // Get value of B
        public double GetC() => _variable3;  // Get value of C
        public double GetD() => _variable4;  // Get value of D
        public double GetExpressionValue()  // Calculate expression from the task
        {
            double expressionValue;
            try
            {
                expressionValue = Math.Log10(4 * _variable2 - _variable3) * _variable1/(_variable2 + _variable3 / _variable4 - 1);
            }
            catch (DivideByZeroException)
            {
                throw new DivideByZeroException("Cannot divide by zero");
            }
            return expressionValue;
        }
    }
}

File: laba6/CPP/Expression/Expression.cpp
#include "Expression.h"

Expression::Expression(double a, double b, double c, double d) {
    this->variable1 = a;
    if (d != 0){
        if (4 * b - c > 0){
            if (b + c / d - 1 != 0){
                this->variable2 = b;
                this->variable3 = c;
                this->variable4 = d;
            } else {
                throw "Wrong value(s) of b/c/d";
            } } else {
            throw "Wrong value(s) of b and/or c";
        } } else {
        throw "Wrong value of d";
    }
}

double Expression::GetA() {
    return this->variable1;
}
double Expression::GetB() {
    return this->variable2;
}
double Expression::GetC() {
    return this->variable3;
}
double Expression::GetD() {
    return this->variable4;
}

double Expression::GetExpressionValue() {
    if (this->variable4 == 0 || (this->variable2 + this->variable3/this->variable4 -1) == 0) {
        throw "Cannot divide by zero";
    }
    return log10(4 * this->variable2 - this->variable3) * this->variable1/(this->variable2 + this->variable3/this->variable4 - 1);
}
